Add coding-guidelines and code-organization files (#11529)
eclipse-theia/theia: doc/coding-guidelines.md

# Coding Guidelines

## Indentation
Use 4 spaces per indentation

## Imports
Use `organize imports` to sort imports, but make sure the imports work properly (i.e. imports from `/src/` for *.ts files are bad but sometimes inserted).

## Names
<a name="pascalcase-type"></a>
* [1.](#pascalcase-type) Use PascalCase for `type` names
<a name="pascalcase-enum"></a>
* [2.](#pascalcase-enum) Use PascalCase for `enum` values
<a name="camelcase-fn"></a>
* [3.](#camelcase-fn)  Use camelCase for `function` and `method` names
<a name="camelcase-var"></a>
* [4.](#camelcase-var)  Use camelCase for `property` names and `local variables`
<a name="whole-words-names"></a>
* [5.](#whole-words-names) Use whole words in names when possible
<a name="lower-case-names"></a>
```ts
// bad
const termWdgId = 1;

// good
const terminalWidgetId = 1;

> ```

Reviewer: This just moves the `<a>` tag down. In the current version it's above the example that belongs to the previous `<a>` tag when it should be below it. I'm not sure why GitHub is showing the whole section changing.

* [6.](#lower-case-names) Use lower case, dash-separated file names (e.g. `document-provider.ts`)
<a name="file-name"></a>
* [7.](#file-name) Name files after the main Type it exports.
> Why? It should be easy to find a type by a file name.
<a name="one-large-class-per-file"></a>
* [7.1](#one-large-class-per-file) Avoid one file with many large classes, put each class in own file. 
> Why? It should be easy to find a class by a file name.
<a name="unique-names"></a>
* [8.](#unique-names) Give unique names to types and files. Use specific names to achieve it.
> Why? In order to avoid duplicate records in file and type search.
```ts
// bad
export interface TitleButton {}

// good
export interface QuickInputTitleButton {}
```
<a name="no_underscore_private"></a>
* [9.](#no_underscore_private) Do not use "_" as a prefix for private properties, exceptions:
<a name="underscore_accessors"></a>
  * [9.1](#underscore_accessors) you want to expose a property through get/set and use underscore for the internal field
<a name="underscore_json"></a>
  * [9.2](#underscore_json) you need to attach internal data to user visible JSON objects
<a name="event_names"></a>

> * [10.](#event_names) Names of events follow the `on[Will|Did]VerbNoun?` pattern. The name signals if the event is going to happen (onWill) or already happened (onDid), what happened (verb), and the context (noun) unless obvious from the context.

Author: I find the original version more concrete on when to add the context (when it's not obvious).

> * [10.](#event_names) Names of events follow the `on[Will|Did]VerbNoun?` pattern. The name signals if the event is going to happen (onWill) or already happened (onDid), what happened (verb), and the context (noun) unless obvious from the context.

Reviewer: Fair enough. I just don't like very much the phrasing that one should add 'the _context_, unless obvious from _context_,' but I couldn't come up with an equally clear formulation.

* [5.](#no-multi-inject) Don't use multi inject, use `ContributionProvider` to inject multiple instances. 
> Why?
> - `ContributionProvider` is a documented way to introduce contribution points. See `Contribution-Points`: https://www.theia-ide.org/docs/services_and_contributions
> - If there is no a single binding, multi inject resolves to `undefined`, not an empty array. `ContributionProvider` provides an empty array.
> - Multi inject does not guarantee the same instances are injected if an extender does not use `inSingletonScope`. `ContributionProvider` caches instances to ensure uniqueness. 

## CSS
<a name="css-use-lower-case-with-dashes"></a>
* [1.](#css-use-lower-case-with-dashes) Use the `lower-case-with-dashes` format.
<a name="css-prefix-global-classes"></a>
* [2.](#css-prefix-global-classes) Prefix classes with `theia` when used as global classes.
<a name="no-styles-in-code"></a>
* [3.](#no-styles-in-code) Do not define styles in code. Introduce proper CSS classes.
> Why? It is not possible to play with such styles in the dev tools without recompiling the code. CSS classes can be edited in the dev tools.

## Theming
<a name="theming-no-css-color-variables"></a>

> * [1.](#theming-no-css-color-variables) Do not introduce css color variables. Implement `ColorContribution` and use `ColorRegistry.register` to register new colors.

Reviewer: ```suggestion
* [1.](#theming-no-css-color-variables) Do not introduce CSS color variables. Implement `ColorContribution` and use `ColorRegistry.register` to register new colors.
```

<a name="theming-no-css-color-values"></a>

> * [2.](#theming-no-css-color-values) Do not introduce hardcode color values in css. Instead reference [VS Code colors](https://code.visualstudio.com/api/references/theme-color) in css by prefixing them with `--theia` and replacing all dots with dashes. For example `widget.shadow` color can be referenced in css with `var(--theia-widget-shadow)`.

Reviewer: ```suggestion
* [2.](#theming-no-css-color-values) Do not introduce hardcoded color values in CSS. Instead refer to [VS Code colors](https://code.visualstudio.com/api/references/theme-color) in CSS by prefixing them with `--theia` and replacing all dots with dashes. For example `widget.shadow` color can be referred to in CSS with `var(--theia-widget-shadow)`.
```

<a name="theming-derive-colors-from-vscode"></a>

> * [3.](#theming-derive-colors-from-vscode) Always derive new colors from existing [VS Code colors](https://code.visualstudio.com/api/references/theme-color). One can derive from an existing color, just by plainly referencing it, e.g. `dark: 'widget.shadow'`, or applying transformations, e.g. `dark: Color.lighten('widget.shadow', 0.4)`. 

Reviewer: ```suggestion
* [3.](#theming-derive-colors-from-vscode) Always derive new colors from existing [VS Code colors](https://code.visualstudio.com/api/references/theme-color). New colors can be derived from an existing color by plain reference, e.g. `dark: 'widget.shadow'`, or transformation, e.g. `dark: Color.lighten('widget.shadow', 0.4)`. 
```

> > Why? Otherwise, there is no guarantee that new colors don't look alien for a random VS Code color theme.

Reviewer: ```suggestion
> Why? Otherwise, there is no guarantee that new colors will fit well into new VSCode color themes.
```

<a name="theming-theia-colors"></a>
* [4.](#theming-theia-colors) Apply different color values only in concrete Theia themes, see [Light (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/light_theia.json), [Dark (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/dark_theia.json) and [High Contrast (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/hc_theia.json) themes.
<a name="theming-variable-naming"></a>
* [5.](#theming-variable-naming) Names of variable follow the `object.property` pattern.
```ts
// bad
'button.secondary.foreground'
'button.secondary.disabled.foreground'

// good
'secondaryButton.foregrond'
'secondaryButton.disabledForegrond'

> ```

Reviewer: @msujew, this section should perhaps be revisited in light of your recent improvements to the theme system?

Author: Let's keep this PR about getting the doc into the repo: I have other edits I want to make after this one lands, as well.

## React
<a name="no-bind-fn-in-event-handlers"></a>
* [1.](#no-bind-fn-in-event-handlers) Do not bind functions in event handlers.

>   - Extract a react component if you want to pass state to an event handler function.

Reviewer: ```suggestion
  - Extract a React component if you want to pass state to an event handler function.
```

> > Why? Because it creates a new instance of event handler function on each render and breaks React element caching leading to rerendering and bad performance.

Reviewer: ```suggestion
> Why? Because doing so creates a new instance of the event handler function on each render and breaks React element caching leading to rerendering and bad performance.
```

```ts
// bad
class MyWidget extends ReactWidget {
  render(): React.ReactNode {
    return <div onClick={this.onClickDiv.bind(this)} />;
  }

  protected onClickDiv(): void {
    // do stuff
  }
}

// bad
class MyWidget extends ReactWidget {
  render(): React.ReactNode {
    return <div onClick={() => this.onClickDiv()} />;
  }

  protected onClickDiv(): void {
    // do stuff
  }
}

// very bad
class MyWidget extends ReactWidget {
  render(): React.ReactNode {
    return <div onClick={this.onClickDiv} />;
  }

  protected onClickDiv(): void {
    // do stuff, no `this` access
  }
}

// good
class MyWidget extends ReactWidget {
  render(): React.ReactNode {
    return <div onClick={this.onClickDiv} />
  }

  protected onClickDiv = () => {
    // do stuff, can access `this`
  }
}
```

## URI/Path
<a name="uri-over-path"></a>
* [1.](#uri-over-path) Pass URIs between frontend and backend, never paths. URIs should be sent as strings in JSON-RPC services, e.g. `RemoteFileSystemServer` accepts strings, not URIs.

> > Why? Frontend and backend can have different operating systems leading to incompatibilities between paths. URIs are normalized in order to be os-agnostic.

Reviewer: ```suggestion
> Why? Frontend and backend can have different operating systems leading to incompatibilities between paths. URIs are normalized in order to be OS-agnostic.
```

<a name="frontend-fs-path"></a>
* [2.](#frontend-fs-path) Use `FileService.fsPath` to get a path on the frontend from a URI.
<a name="backend-fs-path"></a>

> * [3.](#backend-fs-path) Use `FileUri.fsPath` to get a path on the backend from a URI. Never use it on the frontend.

Reviewer: @msujew, these items also fall into your expertise - what's your take on this advice?

<a name="uri-scheme"></a>
* [4.](#explicit-uri-scheme) Always define an explicit scheme for a URI. 

> > Why? A URI without scheme will fall back to `file` scheme for now, in the future it will lead to a runtime error.

Reviewer: ```suggestion
> Why? A URI without scheme will fall back to `file` scheme for now; in the future it will lead to a runtime error.
```

<a name="frontend-path"></a>
* [5.](#frontend-path) Use `Path` Theia API to manipulate paths on the frontend. Don't use Node.js APIs like `path` module. Also see [the code organization guideline](code-organization.md).
<a name="backend-fs"></a>
* [6.](#backend-fs) On the backend, use Node.js APIS to manipulate the file system, like `fs` and `fs-extra` modules.
> Why? `FileService` is to expose file system capabilities to the frontend only. It's aligned with expectations and requirements on the frontend. Using it on the backend is not possible.
<a name="use-long-name"></a>

> * [7.](#use-long-name) Use `LabelProvider.getLongName(uri)` to get a systemwide human readable representation of a full path. Don't use `uri.toString()` or `uri.path.toString()`.

Reviewer: ```suggestion
* [7.](#use-long-name) Use `LabelProvider.getLongName(uri)` to get a system-wide human-readable representation of a full path. Don't use `uri.toString()` or `uri.path.toString()`.
```

<a name="use-short-name"></a>

> * [8.](#use-short-name) Use `LabelProvider.getName(uri)` to get a systemwide human readable representation of a simple file name.

Reviewer: ```suggestion
* [8.](#use-short-name) Use `LabelProvider.getName(uri)` to get a system-wide human-readable representation of a simple file name.
```

<a name="use-icon"></a>

> * [9.](#use-icon) Use `LabelProvider.getIcon(uri)` to get a systemwide file icon.

Reviewer: ```suggestion
* [9.](#use-icon) Use `LabelProvider.getIcon(uri)` to get a system-wide file icon.
```

<a name="uri-no-string-manipulation"></a>
* [10.](#uri-no-string-manipulations) Don't use `string` to manipulate URIs and paths. Use `URI` and `Path` capabilities instead, like `join`, `resolve` and `relative`.
> Why? Because object representation can handle corner cases properly, like trailing separators.
```ts
// bad
uriString + '/' + pathString

// good
new URI(uriString).join(pathString)

// bad
pathString.substring(absolutePathString.length + 1)

// good
new Path(absolutePathString).relative(pathString)
```

## Logging
<a name="logging-use-console-log"></a>

> * [1.](#logging-use-console-log) Use `console` instead of `Ilogger` for the root (top-level) logging.

Reviewer: ```suggestion
* [1.](#logging-use-console-log) Use `console` instead of `ILogger` for the root (top-level) logging.
```

```ts
// bad
@inject(ILogger)
protected readonly logger: ILogger;

this.logger.info(``);

// good
console.info(``)
```

> > Why? All calls to console are intercepted on the frontend and backend, and then forwarded to `ILogger` already. Log level can be configured by `theia start --log-level=debug`.

Reviewer: ```suggestion
> Why? All calls to console are intercepted on the frontend and backend and then forwarded to an `ILogger` instance already. The log level can be configured from the CLI: `theia start --log-level=debug`.
```
